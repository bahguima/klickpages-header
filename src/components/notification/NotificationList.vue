<template>
  <vue-custom-scrollbar
    class="drop__list drop__list--notifications dropdown-menu"
    :tagname="'ul'"
    @ps-y-reach-end="onInfiniteScroll"
  >
    <notification-item
      v-for="notification in notifications.items"
      :key="`${notification.id}`"
      :notification="notification"
    />
    <notification-error
      v-if="showNotificationError"
      @refreshNotifications="refreshNotifications"
      :errorMessage="errorMessage"
    />
    <loader v-if="isLoading" />
  </vue-custom-scrollbar>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';
import VueCustomScrollbar from 'vue-custom-scrollbar';

import NotificationItem from './NotificationItem.vue';
import NotificationError from './NotificationError.vue';
import Loader from './Loader.vue';

const notificationStatus = {
  success: 'success',
  empty: 'empty',
  error: 'error',
};

const ITEMS_PER_PAGE = 5;

export default {
  props: {
    notificationsOpen: Boolean,
  },
  data() {
    return {
      currentNotificationsStatus: notificationStatus.empty,
      currentPage: 1,
      isLoading: false,
    };
  },
  computed: {
    ...mapGetters({
      notifications: 'notification/notifications',
    }),
    emptyNotifications() {
      return this.notifications.length === 0;
    },
    showNotificationError() {
      if (this.isLoading) {
        return false;
      }

      return this.currentNotificationsStatus === notificationStatus.empty
        || this.currentNotificationsStatus === notificationStatus.error;
    },
    errorMessage() {
      if (this.currentNotificationsStatus === notificationStatus.empty) {
        return this.$t('notifications.noNotifications');
      }

      if (this.currentNotificationsStatus === notificationStatus.error) {
        return this.$t('notifications.unableLoad');
      }

      return '';
    },
    notificationsTotalPage() {
      return Math.ceil((this.notifications.totalItems / ITEMS_PER_PAGE));
    },
    shouldLoadMoreNotifications() {
      return this.currentPage < this.notificationsTotalPage;
    },
  },
  components: {
    NotificationItem,
    NotificationError,
    VueCustomScrollbar,
    Loader,
  },
  watch: {
    async notificationsOpen(newValue) {
      if (!newValue) {
        return;
      }

      await this.sendReadNotification();
    },
  },
  methods: {
    ...mapActions({
      getNotifications: 'notification/getNotifications',
      readNotifications: 'notification/readNotifications',
    }),
    async refreshNotifications() {
      try {
        this.isLoading = true;

        await this.getNotifications(this.currentPage);
        this.currentNotificationsStatus = this.emptyNotifications
          ? notificationStatus.empty
          : notificationStatus.success;
      } catch (error) {
        console.error(error);
        this.currentNotificationsStatus = notificationStatus.error;
      } finally {
        this.isLoading = false;
      }
    },
    async sendReadNotification() {
      await this.readNotifications(this.notifications.items);
      this.$emit('notificationsReaded');
    },
    async onInfiniteScroll() {
      if (this.shouldLoadMoreNotifications && !this.isLoading) {
        this.currentPage += 1;
        await this.refreshNotifications();

        if (this.notificationsOpen) {
          this.sendReadNotification();
        }
      }
    },
  },
  async mounted() {
    await this.refreshNotifications();
  },
};
</script>
