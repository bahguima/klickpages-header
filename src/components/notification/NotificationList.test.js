import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import NotificationList from './NotificationList.vue';

jest.mock('@/i18n', () => ({
  t: jest.fn((str) => str),
}));

const $t = jest.fn().mockImplementation((msg) => msg);

console.error = jest.fn();

describe('components/notification/NotificationList', () => {
  const notifications = [
    {
      id: '1',
      data: {
        ptBR: {
          title: 'Teste',
          description: 'notificacao de teste',
          link: '',
          linkText: '',
        },
        en: {
          title: 'Test',
          description: 'test notification',
          link: '',
          linkText: '',
        },
        es: {
          title: 'Prueba',
          description: 'notificación de prueba',
          link: '',
          linkText: '',
        },
      },
      created_at: '12/05/2020',
    },
    {
      id: '2',
      data: {
        ptBR: {
          title: 'Teste com Link',
          description: 'notificacao de teste com Link',
          link: 'https://jwt.io/introduction/',
          linkText: 'json.io',
        },
        en: {
          title: 'Test with link',
          description: 'test notification with link',
          link: 'https://jwt.io/introduction/',
          linkText: 'json.io',
        },
        es: {
          title: 'Prueba',
          description: 'notificación de prueba con enlace',
          link: 'https://jwt.io/introduction/',
          linkText: 'json.io',
        },
      },
      created_at: '12/05/2020',
    },
  ];

  const notification = {
    namespaced: true,
    getters: {
      notifications: jest.fn(() => notifications),
    },
    actions: {
      getNotifications: jest.fn()
        .mockImplementation(() => Promise.resolve()),
    },
  };
  const localVue = createLocalVue();
  localVue.use(Vuex);

  let wrapper;

  describe('when components is mounted', () => {
    describe('and there is notifications', () => {
      beforeAll(() => {
        const store = new Vuex.Store({ modules: { notification } });
        wrapper = shallowMount(NotificationList, { store, localVue, mocks: { $t } });
      });

      afterAll(() => {
        wrapper.destroy();
      });

      it('should render html correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('and there is no notifications', () => {
      beforeAll(() => {
        notification.getters.notifications.mockImplementationOnce(() => []);
        const store = new Vuex.Store({ modules: { notification } });
        wrapper = shallowMount(NotificationList, { store, localVue, mocks: { $t } });
      });

      afterAll(() => {
        wrapper.destroy();
      });

      it('should render html correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('and there occurs an error while trying to retrieve the notifications', () => {
      const mockError = 'mockError';

      beforeAll(() => {
        notification.getters.notifications.mockImplementationOnce(() => []);
        notification.actions.getNotifications
          .mockImplementationOnce(() => Promise.reject(mockError));
        const store = new Vuex.Store({ modules: { notification } });
        wrapper = shallowMount(NotificationList, { store, localVue, mocks: { $t } });
      });

      afterAll(() => {
        wrapper.destroy();
      });

      it('should render html correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('should log the error on the console', () => {
        expect(console.error).toHaveBeenCalledWith(mockError);
      });
    });
  });

  describe('computed', () => {
    describe('emptyNotifications', () => {
      describe('when there is no notification', () => {
        beforeAll(() => {
          notification.getters.notifications.mockImplementationOnce(() => []);
          const store = new Vuex.Store({ modules: { notification } });
          wrapper = shallowMount(NotificationList, { store, localVue, mocks: { $t } });
        });

        afterAll(() => {
          wrapper.destroy();
        });

        it('should return true', () => {
          expect(wrapper.vm.emptyNotifications).toBeTruthy();
        });
      });

      describe('when there is notifications', () => {
        beforeAll(() => {
          const store = new Vuex.Store({ modules: { notification } });
          wrapper = shallowMount(NotificationList, { store, localVue, mocks: { $t } });
        });

        afterAll(() => {
          wrapper.destroy();
        });

        it('should return false', () => {
          expect(wrapper.vm.emptyNotifications).toBeFalsy();
        });
      });
    });
  });
});
