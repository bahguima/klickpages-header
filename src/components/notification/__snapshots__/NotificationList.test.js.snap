// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/notification/NotificationList when components is mounted and there is no notifications should render html correctly 1`] = `
<ul
  class="drop__list drop__list--notifications dropdown-menu"
>
   
  <notification-error-stub
    errormessage="notifications.noNotifications"
  />
</ul>
`;

exports[`components/notification/NotificationList when components is mounted and there is notifications should render html correctly 1`] = `
<ul
  class="drop__list drop__list--notifications dropdown-menu"
>
  <notification-item-stub
    notification="[object Object]"
  />
  <notification-item-stub
    notification="[object Object]"
  />
   
  <!---->
</ul>
`;

exports[`components/notification/NotificationList when components is mounted and there occurs an error while trying to retrieve the notifications should render html correctly 1`] = `
<ul
  class="drop__list drop__list--notifications dropdown-menu"
>
   
  <notification-error-stub
    errormessage="notifications.unableLoad"
  />
</ul>
`;
